File: 0151-reverse-words-in-a-string/0151-reverse-words-in-a-string.java
class Solution {
    private void reverseCharArray(char[] chars, int start, int end) {
        while (start < end) {
            char tmp = chars[start];
            chars[start] = chars[end];
            chars[end] = tmp;
            start++;
            end--;
        }
    }

    public String reverseWords(String s) {
        char[] chars = s.toCharArray();
        int n = chars.length;

        reverseCharArray(chars, 0, n - 1);

        int wordEnd = 0;
        for (int i = 0; i < n;) {
            while (i < n && chars[i] == ' ') i++;

            if (i == n) break;

            if (wordEnd > 0) chars[wordEnd++] = ' ';

            int wordStart = wordEnd;

            while (i < n && chars[i] != ' ') {
                chars[wordEnd++] = chars[i++];
            }

            reverseCharArray(chars, wordStart, wordEnd - 1);
        }

        return new String(chars, 0, wordEnd);
    }
}

File: 0151-reverse-words-in-a-string/0151-reverse-words-in-a-string.kt
class Solution {
    private fun reverseCharArray(chars: CharArray, start: Int, end: Int) {
        var left = start
        var right = end

        while (left < right) {
            val tmp = chars[left]
            chars[left] = chars[right]
            chars[right] = tmp
            left++
            right--
        }
    }

    fun reverseWords(s: String): String {
        val chars = s.toCharArray()
        val n = chars.size

        reverseCharArray(chars, 0, n - 1)

        var wordEnd = 0
        var i = 0
        while (i < n) {
            while (i < n && chars[i] == ' ') i++

            if (i == n) break

            if (wordEnd > 0) chars[wordEnd++] = ' '

            val wordStart = wordEnd

            while (i < n && chars[i] != ' ') {
                chars[wordEnd++] = chars[i++]
            }

            reverseCharArray(chars, wordStart, wordEnd - 1)
        }

        return String(chars, 0, wordEnd)
    }
}